src/test/java/org/jenkinsci/plugins/workflow/steps/RetryStepTest.java

```diff
@@ -0,0 +1,47 @@
+package org.jenkinsci.plugins.workflow.steps;
+
+import hudson.model.Result;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.test.steps.SemaphoreStep;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import java.io.IOException;
+
+import static org.junit.Assert.*;
+
+/**
+ * Tests {@link RetryStep}.
+ */
+public class RetryStepTest {
+
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @Test(timeout = 50000)
```

Comment:
Worse, the test _fails_ when the `timeout` is removed. I think because of a race condition: the call to `recordCauseOfInterruption` from `WorkflowRun` happens asynchronously. So the `InterruptedBuildAction` may or may not be present at the time the `onFailure` handler is called, according to timing. So the current patch is doomed.